sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuFileFormatDataWriter.scala

```diff
@@ -163,28 +165,43 @@ class GpuSingleDirectoryDataWriter(
   }
 
   override def write(batch: ColumnarBatch): Unit = {
-    var needToCloseBatch = true
-    try {
-      // TODO: This does not handle batch splitting to make sure maxRecordsPerFile is not exceeded.
-      if (description.maxRecordsPerFile > 0 && recordsInFile >= description.maxRecordsPerFile) {
-        fileCounter += 1
-        assert(fileCounter < MAX_FILE_COUNTER,
-          s"File counter $fileCounter is beyond max value $MAX_FILE_COUNTER")
-
-        newOutputWriter()
-      }
-
+    val maxRecordsPerFile = description.maxRecordsPerFile
+    if (!needSplitBatch(maxRecordsPerFile, recordsInFile, batch.numRows())) {
       statsTrackers.foreach(_.newBatch(batch))
-      recordsInFile += batch.numRows
-      needToCloseBatch = false
-    } finally {
-      if (needToCloseBatch) {
-        batch.close()
+      recordsInFile += batch.numRows()
+      currentWriter.write(batch, statsTrackers)
+    } else {
+      withResource(batch) { batch =>
+        withResource(GpuColumnVector.from(batch)) {table => 
+          val splitIndexes = getSplitIndexes(
+            maxRecordsPerFile,
+            recordsInFile,
+            table.getRowCount()
+          )
+  
+          val dataTypes = (0 until batch.numCols()).map(i => batch.column(i).dataType()).toArray
+  
+          var needNewWriter = recordsInFile >= maxRecordsPerFile && maxRecordsPerFile > 0
```

Comment:
`maxRecordsPerFile > 0` is no need.
And `recordsInFile >= maxRecordsPerFile` here actually want to test `recordsInFile == maxRecordsPerFile`,  recordsInFile should not be greater than maxRecordsPerFile. I think we can change to:
assert(recordsInFile <= maxRecordsPerFile)
var needNewWriter = recordsInFile == maxRecordsPerFile